frame/support/src/dispatch.rs

```diff
@@ -1302,42 +1303,52 @@ macro_rules! decl_module {
 			for $call_type<$trait_instance $(, $instance)?> where $( $other_where_bounds )*
 		{
 			fn get_dispatch_info(&self) -> $crate::dispatch::DispatchInfo {
-				$(
-					if let $call_type::$fn_name($( ref $param_name ),*) = self {
-						let weight = <dyn $crate::dispatch::WeighData<( $( & $param, )* )>>::weigh_data(
-							&$weight,
-							($( $param_name, )*)
-						);
-						let class = <dyn $crate::dispatch::ClassifyDispatch<( $( & $param, )* )>>::classify_dispatch(
-							&$weight,
-							($( $param_name, )*)
-						);
-						let pays_fee = <dyn $crate::dispatch::PaysFee<( $( & $param, )* )>>::pays_fee(
-							&$weight,
-							($( $param_name, )*)
-						);
-						return $crate::dispatch::DispatchInfo { weight, class, pays_fee };
-					}
-					if let $call_type::__PhantomItem(_, _) = self { unreachable!("__PhantomItem should never be used.") }
-				)*
-				// Defensive only: this function must have already returned at this point.
-				// all dispatchable function will have a weight which has the `::default`
-				// implementation of `SimpleDispatchInfo`. Nonetheless, we create one if it does
-				// not exist.
-				let weight = <dyn $crate::dispatch::WeighData<_>>::weigh_data(
-					&$crate::dispatch::SimpleDispatchInfo::default(),
-					()
-				);
-				let class = <dyn $crate::dispatch::ClassifyDispatch<_>>::classify_dispatch(
-					&$crate::dispatch::SimpleDispatchInfo::default(),
-					()
-				);
-				let pays_fee = <dyn $crate::dispatch::PaysFee<_>>::pays_fee(
-					&$crate::dispatch::SimpleDispatchInfo::default(),
-					()
-				);
-				$crate::dispatch::DispatchInfo { weight, class, pays_fee }
+				match *self {
+					$(
+						$call_type::$fn_name( $( ref $param_name ),* ) => {
+							let weight = <dyn $crate::dispatch::WeighData<( $( & $param, )* )>>::weigh_data(
+								&$weight,
+								($( $param_name, )*)
+							);
+							let class = <dyn $crate::dispatch::ClassifyDispatch<( $( & $param, )* )>>::classify_dispatch(
+								&$weight,
+								($( $param_name, )*)
+							);
+							let pays_fee = <dyn $crate::dispatch::PaysFee<( $( & $param, )* )>>::pays_fee(
+								&$weight,
+								($( $param_name, )*)
+							);
+							$crate::dispatch::DispatchInfo { 
+								weight, 
+								class, 
+								pays_fee,
+							}
+						},
+					)*
+					_ => unreachable!("__PhantomItem should never be used."),
```

Comment:
If the last variant is only the phantom item then we should be able to match it directly `$call_type::__PhantomItem(_, _)`